code/cert/cert_pkcs.c

```diff
@@ -140,10 +140,22 @@ BYTE * cert_pkcs_sign(struct ssh2_userkey * userkey, LPCBYTE pDataToSign, int iD
 	// check for error
 	if (hSession == 0 || hPublicKey == 0)
 	{
-		// error
-		return NULL;
-	}
-
+		// when public key don�t exist(after import p12 in SafeNet 5100),
+		// we try to locate key id by certificate�s id
+		CK_BBOOL bFalse = CK_FALSE;
+		CK_BBOOL bTrue = CK_TRUE;
+		CK_OBJECT_CLASS iObjectType = CKO_CERTIFICATE;
+		CK_ATTRIBUTE aFindCriteria[] = {
+			{ CKA_CLASS,    &iObjectType, sizeof(CK_OBJECT_CLASS) },
+			{ CKA_TOKEN,    &bTrue,       sizeof(CK_BBOOL) },
+			{ CKA_PRIVATE,  &bFalse,      sizeof(CK_BBOOL) }
+		};
+		pkcs_lookup_token_cert(userkey->comment, &hSession, &hPublicKey,
+			aFindCriteria, _countof(aFindCriteria), FALSE);
+		if (hSession == 0 || hPublicKey == 0)
+			// error
+			return NULL;
+	} 
```

Comment:
Can you provide some background here?  Is this a problem we're trying to workaround?

Reply:
Usually, here, we search public key to get ID (CKA_ID) which let to find private key. The problem is what in some cases public key doesn’t exist - only certificate which have the same ID as private key. So, in this code we replace handle of non-existing public key by existing handle of certificate and use this handle to read ID of private key.
Please note what in common case one private key has 2 public keys – public key him-self and public key inside certificate. This duplication reduces the number of keys which can be stored inside a token.  For example, with token SafeNet 5100 and key RSA2048, we lose 7 pairs private key and certificate to store 11 public keys.  In other words, one token can contain 18 keys without public key and 11 keys with public key.

Reply:
I adjusted your changes for the comment generalization and cleanup.  Look good to you?  If so, I'll merge it.

Reply:
It is very good for me
thank you